api/v1/storageprofile_types.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2020 Red Hat OpenShift Container Storage.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StorageProfileSpec defines the desired state of StorageProfile
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.deviceClass) || has(self.deviceClass)", message="deviceClass is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.sharedFilesystemConfiguration) || has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
+type StorageProfileSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="DeviceClass is immutable"
+	// +kubebuilder:validation:MaxLength=512
+	// DeviceClass is the deviceclass name.
+	DeviceClass string `json:"deviceClass"`
```

Comment:
I followed the guide [here](https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/#immutablility-after-first-modification):

> To enforce a field's immutability after its first write, you need to apply the following constraints:
Field must be allowed to be initially unset `+kubebuilder:validation:Optional`

Reply:
@raaizik So, in the case where a storage class was created without a device set, can you please answer the following questions:

- what will be the behavior of a storage profile reconciliation? 
- which device class will be used? 
- Are we going to reject this storage profile? 
- If we are rejecting, are we going to reflect that in the status section of the profile?

Reply:
Currently this kind of `StorageProfile` will be appended with no `deviceClass`
```
% oc get storageprofile
NAME                                AGE
empty-performance-storageprofile    38m
fast-performance-storageprofile     100m
medium-performance-storageprofile   40m
slow-performance-storageprofile     40m
``` 
`empty-performance-storageprofile`'s spec:
```
Spec:
  Block Pool Configuration:
    Parameters:
      pg_autoscale_mode:  on
      pg_num:             128
      pgp_num:            128
  Shared Filesystem Configuration:
    Parameters:
      pg_autoscale_mode:  on
      pg_num:             128
      pgp_num:            128
Events:                   <none>
``` 
```
Spec:
  Data Pools:
    Device Class:  fast
    Erasure Coded:
      Coding Chunks:  0
      Data Chunks:    0
    Failure Domain:   zone
    Mirroring:
    Name:  fast
    Parameters:
      pg_autoscale_mode:  on
      pg_num:             128
      pgp_num:            128
    Quotas:
    Replicated:
      Replicas Per Failure Domain:  1
      Size:                         3
      Target Size Ratio:            0.49
    Status Check:
      Mirror:
    Device Class:  medium
    Erasure Coded:
      Coding Chunks:  0
      Data Chunks:    0
    Failure Domain:   zone
    Mirroring:
    Name:  medium
    Parameters:
      pg_autoscale_mode:  on
      pg_num:             128
      pgp_num:            128
    Quotas:
    Replicated:
      Replicas Per Failure Domain:  1
      Size:                         3
      Target Size Ratio:            0.49
    Status Check:
      Mirror:
    Device Class:  slow
    Erasure Coded:
      Coding Chunks:  0
      Data Chunks:    0
    Failure Domain:   zone
    Mirroring:
    Name:  slow
    Parameters:
      pg_autoscale_mode:  on
      pg_num:             128
      pgp_num:            128
    Quotas:
    Replicated:
      Replicas Per Failure Domain:  1
      Size:                         3
      Target Size Ratio:            0.49
    Status Check:
      Mirror:
    Erasure Coded:
      Coding Chunks:  0
      Data Chunks:    0
    Failure Domain:   zone
    Mirroring:
    Parameters:
      pg_autoscale_mode:  on
      pg_num:             128
      pgp_num:            128
``` 

Once a `deviceClass` is set, the reconciliation appears to fail:
```
  Warning  ReconcileFailed     28s                    rook-ceph-file-controller  failed to reconcile CephFilesystem "openshift-storage/ocs-storagecluster-cephfilesystem". failed to create filesystem "ocs-storagecluster-cephfilesystem": failed to create filesystem "ocs-storagecluster-cephfilesystem": failed to set pools size: failed to update datapool  "ocs-storagecluster-cephfilesystem-empty": failed to create replicated crush rule "ocs-storagecluster-cephfilesystem-empty": failed to create crush rule ocs-storagecluster-cephfilesystem-empty: exit status 22
```

Reply:
Ok, so the plan is to just reject this kind of storage profile and not take it into consideration when new requests come up. correct?

Reply:
Note that the status is set to blank in case of then setting the `deviceClass`, which pretty much leaves the two other phases unused. IMO not the most optimal thing to have, but I did not see spots where such updates are required.